File: courtlistener/api/alerts.py
"""
Alerts API module for CourtListener SDK.
"""

from typing import Dict, Any, Optional, List
from .base import BaseAPI


class AlertsAPI(BaseAPI):
    """API for managing search and docket alerts."""
    
    def __init__(self, client):
        super().__init__(client)
        self.endpoint = "alerts/"
    
    def list(
        self,
        name: Optional[str] = None,
        alert_type: Optional[str] = None,
        **kwargs
    ) -> Dict[str, Any]:
        """
        List alerts.
        
        Args:
            name: Filter by alert name
            alert_type: Filter by alert type (search, docket)
            **kwargs: Additional query parameters
        
        Returns:
            API response with paginated list of alerts
        """
        params = {}
        if name is not None:
            params['name'] = name
        if alert_type is not None:
            params['alert_type'] = alert_type
        
        params.update(kwargs)
        return self.client.get(self.endpoint, params=params)
    
    def get(self, alert_id: int) -> Dict[str, Any]:
        """
        Get a specific alert by ID.
        
        Args:
            alert_id: Alert ID
        
        Returns:
            Alert data
        """
        return self.client.get(f"{self.endpoint}{alert_id}/")
    
    def create(
        self,
        name: str,
        query: str,
        rate: str,
        alert_type: Optional[str] = None,
        webhook_url: Optional[str] = None,
        **kwargs
    ) -> Dict[str, Any]:
        """
        Create a new alert.
        
        Args:
            name: Name for the alert
            query: Search query string
            rate: Alert frequency (dly, wly, mly)
            alert_type: Type of alert (search, docket)
            webhook_url: Webhook URL for notifications
            **kwargs: Additional parameters
        
        Returns:
            Created alert data
        """
        data = {
            'name': name,
            'query': query,
            'rate': rate
        }
        if alert_type is not None:
            data['alert_type'] = alert_type
        if webhook_url is not None:
            data['webhook_url'] = webhook_url
        
        data.update(kwargs)
        return self.client.post(self.endpoint, data=data)
    
    def update(
        self,
        alert_id: int,
        **kwargs
    ) -> Dict[str, Any]:
        """
        Update an existing alert.
        
        Args:
            alert_id: Alert ID
            **kwargs: Fields to update
        
        Returns:
            Updated alert data
        """
        return self.client.post(f"{self.endpoint}{alert_id}/", data=kwargs)
    
    def delete(self, alert_id: int) -> bool:
        """
        Delete an alert.
        
        Args:
            alert_id: Alert ID
        
        Returns:
            True if successful
        """
        self.client._make_request('DELETE', f"{self.endpoint}{alert_id}/")
        return True
    
    def paginate(
        self,
        name: Optional[str] = None,
        alert_type: Optional[str] = None,
        **kwargs
    ):
        """
        Get paginated alerts.
        
        Args:
            name: Filter by alert name
            alert_type: Filter by alert type (search, docket)
            **kwargs: Additional query parameters
        
        Returns:
            PageIterator for alerts
        """
        params = {}
        if name is not None:
            params['name'] = name
        if alert_type is not None:
            params['alert_type'] = alert_type
        
        params.update(kwargs)
        return self.client.paginate(self.endpoint, params=params)

File: php/src/CourtListener/Api/Alerts.php
<?php

namespace CourtListener\Api;

use CourtListener\CourtListenerClient;
use CourtListener\Exceptions\CourtListenerException;

/**
 * Alerts API endpoint
 */
class Alerts extends BaseApi
{
    protected string $endpoint = 'alerts/';

    /**
     * Constructor
     *
     * @param CourtListenerClient $client
     */
    public function __construct(CourtListenerClient $client)
    {
        parent::__construct($client);
    }

    /**
     * List alerts with advanced filtering
     *
     * @param array $params Query parameters
     * @return array
     * @throws CourtListenerException
     */
    public function listAlerts(array $params = []) {
        return $this->list($params);
    }

    /**
     * Get a specific alert by ID
     *
     * @param int|string $id Alert ID
     * @param array $params Query parameters
     * @return array
     * @throws CourtListenerException
     */
    public function getAlert(array $params = []) {
        return $this->get($id, $params);
    }

    /**
     * Create a new alert
     *
     * @param array $data Alert data
     * @return array
     * @throws CourtListenerException
     */
    public function createAlert(array $params = []) {
        return $this->create($data);
    }

    /**
     * Update an existing alert
     *
     * @param int|string $id Alert ID
     * @param array $data Alert data
     * @return array
     * @throws CourtListenerException
     */
    public function updateAlert(array $params = []) {
        return $this->update($id, $data);
    }

    /**
     * Delete an alert
     *
     * @param int|string $id Alert ID
     * @return array
     * @throws CourtListenerException
     */
    public function deleteAlert(array $params = []) {
        return $this->delete($id);
    }

    /**
     * Search alerts
     *
     * @param array $params Search parameters
     * @return array
     * @throws CourtListenerException
     */
    public function searchAlerts(array $params = []) {
        return $this->search($params);
    }
}
